### ISSUE TITLE
Attempting to generate debug info using non-semantic shader debug information results in "Id is 0"

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/glslang#3301]
I am trying to generate shader debug information with slang (https://github.com/shader-slang/slang) which uses glslang under the hood.  It's just been updated to commit 4f3ae4b (https://github.com/KhronosGroup/glslang/commit/8c8c6027937faad014feb583b05f345353d5ed6d) and this bug also reproduces with Vulkan SDK 1.3.246.1. glslangValidator:

```
Glslang Version: 11:12.1.0
ESSL Version: OpenGL ES GLSL 3.20 glslang Khronos. 12.1.0
GLSL Version: 4.60 glslang Khronos. 12.1.0
SPIR-V Version 0x00010600, Revision 1
GLSL.std.450 Version 100, Revision 1
Khronos Tool ID 8
SPIR-V Generator Version 11
GL_KHR_vulkan_glsl version 100
ARB_GL_gl_spirv version 100
```

To reproduce, with the attached shader, run:
```
glslangvalidator -gVS -S frag -Os -V 1.glsl
```

It gives the following error:
```
error: input:0:0:205: Id is 0
```

The error goes away if `-Os` is not used or `-gV[S]`is not used.  So it's a bug with optimization and shader debug info.  It could be spirv-opt related, but thought I'd start here.

Shader attached:

[1.zip](https://github.com/KhronosGroup/glslang/files/12322114/1.zip)

**Comment by danginsburg** [comment]
This appears to be triggered by the `buffer_reference` in the shader, here is a more stripped down version.  If you comment out the buffer_reference the error does not occur:

```
#version 450
#extension GL_EXT_buffer_reference : require
#extension GL_ARB_enhanced_layouts : require

layout(buffer_reference, std430, buffer_reference_align = 16) readonly buffer BufferPointer_BDATest_t_0_1
{
    vec4 _data;
};

layout(binding = 0)
layout(std140) uniform _S2
{
    BufferPointer_BDATest_t_0_1 g_vBDAValue_0;
} ExternalTestCB1_0;

layout(location = 0) out vec4 _S5;


void main()
{
    _S5 = vec4(0,0,0,0);
    return;
}


```

**Comment by arcady-lunarg** [comment]
I'll take a look at this today. Given it's triggered only by -Os, I would suspect this is a SPIRV-Tools bug though.

**Comment by arcady-lunarg** [comment]
Actually, this is a bug in glslang's generation of debuginfo for buffer references, with a debug build and no -Os, I get an assertion failure.
```
glslangValidator: /home/arcady/glslang/SPIRV/SpvBuilder.cpp:990: spv::Id spv::Builder::makeMemberDebugType(const spv::Id, const spv::Builder::DebugTypeLoc &): Assertion `debugId[memberType] != 0' failed.
```

**Comment by arcady-lunarg** [comment]
In fact, I think the actual bug here is that no debug info is emitted for the buffer reference pointer type at all, so when a struct has a buffer reference member, the code that emits debug info for the struct's members tries to find the type of that member, can't find it, and asserts (in debug builds) or emits a 0 (in release builds).

**Comment by greg-lunarg** [comment]
The original author generated SPIR-V for buffer references so that it would work if the buffer reference was recursive or not. So he always generated the pointer type after the struct that contained it.

I am guessing since the pointer type is not generated at the time the struct type is generated, the debug info for the pointer type is not generated at the time the debug info for the struct is being generated.

Since SPIR-V allows forward references in struct types, this policy works when generating the core SPIR-V. But since (currently) there is no way to do forward references for NonSemantic.Shader.DebugInfo, this won't fly.

The shader of this issue does not require forward references for either the core SPIR-V or the NonSemantic.Shader.DebugInfo, so there is no reason the NonSemantic.Shader.DebugInfo for this shader cannot be generated as the specs exist today. We will just have to add extra logic to glslangValidator to do it. Specifically, we will need glslangValidator to generate the debug info for the pointer type before the debug info for the struct type that contains it, when possible.

We still won't be able to generate NonSemantic.Shader.DebugInfo for truly recursive (linked-list) structs until the current plans for the SPIR-V and NonSemantic.Shader.DebugInfo specs https://github.com/KhronosGroup/SPIRV-Registry/issues/203 are completed.

**Comment by RikoOphorst** [comment]
Pinging @jeremy-lunarg as well, I don't know how visible this comment is 😅

**Comment by arcady-lunarg** [comment]
There is going to be a more comprehensive solution to this soon, but it requires a new SPIR-V extension which is currently making its way through the approval process. I am going to take a look at your example to see if it can be fixed easily to not emit a 0 Id with the current scheme, and to make sure that it will work with the new solution with the new extension.

**Comment by RikoOphorst** [comment]
Thanks a lot, I hope it's trivial enough to find a workaround! Also, is there an approximate ETA for that new SPIR-V extension to make it's way out into the wild?

**Comment by arcady-lunarg** [comment]
This is fixed now, the new extension has been released and glslang now uses it as of #3617.
